### PR TITLE
Added CMake Build Support

### PR DESCRIPTION
**Pull request by ambasta**
Hi,

Currently I've added the ability to build and install the C++ component. I've some rudimentary work in place to build the python components as well, though I'll be adding the missing bits in a while.

This PR aims to address #292

### CODE REVIEW

File: ortools/algorithms/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+ENABLE_LANGUAGE(CXX)
+# ENABLE_LANGUAGE(Python)
+SET(ROOT_PROJECT ${PROJECT_NAME})
+PROJECT(${PROJECT_NAME}_algorithms)
+
+FILE(GLOB ${PROJECT_NAME}_SOURCES "*.cc")
+LIST(REMOVE_ITEM ${PROJECT_NAME}_SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/hungarian_test.cc")
```
Comment (lperron) [inline ortools/algorithms/CMakeLists.txt]: replace hungarian_test.cc by *_test.cc

Can you add one test rule per _test.cc?

Comment (ambasta) [inline ortools/algorithms/CMakeLists.txt]: Sure. Currently trying to allow for ExternalProject_Add for mac/windows. Also trying to figure out a way to install the python module. I had actually submitted the PR for review before merge, but I this was really quick. I'll be updating the CMake build over the next few days to address all the issues!

**Comment by lperron** [comment]
Can you tell me how to run cmake?
It complains about missing DIRECTORY?

I think on Mac, I will recommend using homebrew to install dependencies
On windows, it will be useful.

Thanks!

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-30 14:19 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by ambasta** [comment]
mkdir build && cd build && cmake .. -DBUILD_CXX=ON && make && make install

**Comment by lperron** [comment]
CMake Error at CMakeLists.txt:77 (INSTALL):

  install TARGETS given no LIBRARY DESTINATION for shared library target

  "ortools".


??

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-30 14:37 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by ambasta** [comment]
Right, I'm not sure where to install files on non-linux architectures. I'll submit a patch by EoD to set appropriate directories for Windows/Mac

**Comment by lperron** [comment]
I made some inroad

/Applications/CMake.app/Contents/bin/cmake -DBUILD_CXX=ON
-DCMAKE_INSTALL_LIBDIR=/Users/lperron/Work/or-tools-dist/lib
-DCMAKE_INSTALL_BINDIR=/Users/lperron/Work/or-tools-dist/bin
-DCMAKE_INSTALL_INCLUDEDIR=/Users/lperron/Work/or-tools-dist/include ..

finished OK on mac.

But compilation fails:


[  0%] Building CXX object
CMakeFiles/ortoolsProto.dir/ortools/bop/bop_parameters.pb.cc.o

clang: *error: *no input files

/bin/sh: -I/usr/local/Cellar/cgl/0.59.6/include/cgl/coin: No such file or
directory

/bin/sh: -I/usr/local/Cellar/clp/1.16.7/include/clp/coin: No such file or
directory

/bin/sh: -I/usr/local/Cellar/osi/0.107.5/include/osi/coin: No such file or
directory

/bin/sh: -I/usr/local/Cellar/coinutils/2.10.10/include/coinutils/coin: No
such file or directory

/bin/sh: -I/usr/local/Cellar/coinutils/2.10.10/include/coinutils/coin: No
such file or directory

make[2]: ***
[CMakeFiles/ortoolsProto.dir/ortools/bop/bop_parameters.pb.cc.o] Error 127

make[1]: *** [CMakeFiles/ortoolsProto.dir/all] Error 2

make: *** [all] Error 2


Please note that all missing include directories do exist.

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-30 14:47 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by ambasta** [comment]
Hi @lperron 

You're correct about the missing directories existing. However, currently, the cmake project expects the user to have specified the dependencies. I'm enabling ExternalProject_Add in a new set of patches. For now I'm also removing the Python bits so I can work on them separately and not muddle the patches!

**Pull request merged**